flocker/ca/functional/test_validation.py

```diff
@@ -0,0 +1,274 @@
+    ControlServicePolicy, amp_server_context_factory, rest_api_context_factory,
+    )
+
+
+EXPECTED_STRING = b"Mr. Watson, come here; I want to see you."
+
+
+class SendingProtocol(Protocol):
+    """
+    Send a string.
+    """
+    def connectionMade(self):
+        self.transport.write(EXPECTED_STRING)
+        self.transport.loseConnection()
+
+
+class ReceivingProtocol(Protocol):
+    """
+    Expect a string.
+
+    :ivar Deferred result: Fires on receiving response or if disconnected
+         before that.
+    """
+    def __init__(self):
+        self.result = Deferred()
+        self._buffer = b""
+
+    def dataReceived(self, data):
+        self._buffer += data
+        if self._buffer == EXPECTED_STRING:
+            self.result.callback(None)
+            self.result = None
+
+    def connectionLost(self, reason):
+        if self.result:
+            self.result.errback(reason)
+            self.result = None
+
+
+class CredentialSet(PRecord):
+    """
+    A full set of credentials for a CA.
+
+    :param RootCredential root: The CA root credential.
+    :param ControlCredential control: A control service credential.
+    :param UserCredential user: A user credential.
+    :param NodeCredential node: A CA root credentials.
+    """
+    root = field()
+    control = field()
+    user = field()
+    node = field()
+
+    @staticmethod
+    def create():
+        """
+        :return: A new ``CredentialSet``.
+        """
+        directory = FilePath(mkdtemp())
+        atexit.register(rmtree, directory.path)
+        root = RootCredential.initialize(directory, b"mycluster")
+        user = UserCredential.initialize(directory, root, u"allison")
+        node = NodeCredential.initialize(directory, root)
+        control = ControlCredential.initialize(directory, root, b"127.0.0.1")
+        return CredentialSet(root=root, user=user, node=node, control=control)
+
+
+def get_sets():
+    """
+    :return: Tuple, a pair of ``CredentialSet`` instances.
+    """
+    a = CredentialSet.create()
+    b = CredentialSet.create()
+    global get_sets
+
+    def get_sets():
+        return a, b
+    return get_sets()
+
+
+def make_validation_tests(context_factory_fixture,
+                          good_certificate_name,
+                          validator_is_client):
+    """
+    Create a ``TestCase`` for the validator of a specific certificate type.
+
+    :param context_factory_fixture: Create a context factory that
+         implements the required validation given a ``CredentialSet``,
+         given a port number on localhost and a ``CertificateSet``. For
+         server context factories the port number can be ignored.
+
+    :param str good_certificate_name: Name of certificate (an attribute of
+        ``CredentialSet``) that should validate successfully.
+
+    :param bool validator_is_client: Whether or not the context factory
+         being tested is a client.
+
+    :return TestCase: Tests for given validator.
```

Comment:
I think `:return type:` is not actually valid markup.  And the return type here is `type` anyway. :smile: